Allow setting file permissions via API and `lxd file push --mode`

MggMuggins opened the pull request:
Add an API extension `instances_files_modify_permissions` which allows setting the permissions of existing files using the file API. Also updates `lxc` to send the `X-LXD-modify-perm` header when `--mode`, `--uid`, or `--gid` are passed.

I elected to go with a new header rather than adding complexity to the existing headers as discussed in https://github.com/canonical/lxd/issues/12724; maintaining the client's backwards compatibility with older servers would have been non-trivial. This way the new header is simply ignored by older servers.

I'm happy to add integration tests if needed.

Closes https://github.com/canonical/lxd/issues/12724

tomponline commented:
please can you take a look at those integer conversion issues?

tomponline commented:
static analysis has some linter fixes needed too

tomponline commented on shared/version/api.go:
```diff
@@ -401,6 +401,7 @@ var APIExtensions = []string{
 	"container_syscall_filtering_allow_deny_syntax",
 	"access_management",
 	"vm_disk_io_limits",
+	"instances_files_modify_permissions",
```
please add api extension and associated doc change as a separate single commit, e.g. like 

https://github.com/canonical/lxd/commit/5d1ab06926d2cfa9d91324a832a3eca5544fd507

See https://github.com/canonical/lxd/blob/main/CONTRIBUTING.md#commit-structure

tomponline replied:
Also for ease of backporting please make it the first commit in your PR so I can see what is coming up.

tomponline commented on client/lxd_containers.go:
```diff
@@ -809,12 +809,16 @@ func (r *ProtocolLXD) GetContainerFile(containerName string, path string) (io.Re
 
 	// Parse the headers
-	uid, gid, mode, fileType, _ := shared.ParseLXDFileHeaders(resp.Header)
+	headers, err := shared.ParseLXDFileHeaders(resp.Header)
```
Please split up this commit into logical changes - client changes separately from others so it can be backported independent of the server changes.

tomponline commented on shared/util.go:
```diff
@@ -257,18 +257,32 @@ func LogPath(path ...string) string {
 
-func ParseLXDFileHeaders(headers http.Header) (uid int64, gid int64, mode int, type_ string, write string) {
+type LXDFileHeaders struct {
+	Uid               int64
```
I believe go lint will pick you up on using `Uid` and instead should be `UID`. In general acronyms in vars should be all in caps if being exported.

MggMuggins commented:
Fixed some of the linter issues; I'll rebase and fix the rest asap.

I split up the changes based on directory; I'm happy to modify commit messages as needed. I admit some confusion over splitting up the change this way; I broke API in `shared/` by changing the return type of `ParseLXDFileHeaders`, so if someone were to check out any of these commits other than the last one they might end up with build errors.

The unchecked casts that CodeQL is complaining about are currently exposed in the API as implicit integer overflow, e.g. this:
```
lxc file push ../hello.sh test1/home/test/hello.sh --uid 4294967297
```
results in:
```
-rwxr-xr-x 1 daemon ubuntu   36 Mar 22 21:02 hello.sh*
```
where 
```
$ id daemon
uid=1(daemon) gid=1(daemon) groups=1(daemon)
```
I'm adding a commit that ignores `uid` and `gid` values over the 32-bit max and truncates any non-mode bits in the `mode` input. While this is "API breakage", I would argue that this is UB at the moment so breaking it should be alright.

MggMuggins commented on shared/util.go:
```diff
@@ -257,44 +257,95 @@
+			mode64 = rawMode
 		}
 	}
+	mode := int(os.FileMode(mode64) & os.ModePerm)
```
`os.FileMode` is `uint32`, but since we're truncating to `os.ModePerm` anyway I would argue that this is a wontfix.

tomponline replied:
We should try and fix it in a way that the code scanner detects it as fixed.

I believe the codeql issue itself suggests ways to fix it, using condition checks for a valid range.

tomponline commented:
> I split up the changes based on directory; I'm happy to modify commit messages as needed. I admit some confusion over splitting up the change this way; I broke API in `shared/` by changing the return type of `ParseLXDFileHeaders`, so if someone were to check out any of these commits other than the last one they might end up with build errors.

In general we're not too concerned about every commit being buildable. But more that each commit consists of a logical change in order to assist with code review.

MggMuggins commented:
Thanks for that note, I'll keep it in mind for future PRs. I think I've addressed all your feedback, happy to make additional changes if needed.

tomponline commented:
@MggMuggins hey, please could you update the PR description and the API extension description to document the new header format so aid in the code review and future readers.

tomponline commented on shared/util.go:
```diff
@@ -257,18 +257,36 @@ func LogPath(path ...string) string {
+
+// ParseLXDFileHeaders parses and validates the `X-LXD-*` family of file
+// permissions headers.
+func ParseLXDFileHeaders(headers http.Header) (LXDFileHeaders, error) {
```
It would be really useful to the reader for the doc block of this function to document the format of the headers its going to parse.

tomponline commented on shared/util.go:
```diff
@@ -278,23 +296,55 @@ func ParseLXDFileHeaders(headers http.Header) (uid int64, gid int64, mode int, t
+			modifyExistingMode = modifyExistingMode || perm == "mode"
+
+			if !ValueInSlice(perm, []string{"uid", "gid", "mode"}) {
+				return LXDFileHeaders{}, fmt.Errorf("Bad modify-perm field: %q", perm)
```
Can we say "Invalid modify-perm field value:" rather than "Bad ..."

MggMuggins commented:
Added some more docs and fixes as you suggested and rebased to main; not sure why the tests are failing here.

tomponline commented on shared/util_test.go:
```diff
@@ -40,6 +40,87 @@ func TestUrlsJoin(t *testing.T) {
+	}
+
+	if headers.UID != 0 || headers.GID != 99 || headers.Mode != 0o644 {
+		t.Fatalf("Bad Uid (%d), Gid (%d), or Mode (%d)", headers.UID, headers.GID, headers.Mode)
```
In Go its generally expected and linted for that acronyms should be all upper case. So "UID" and "GID" here rather than "Uid" and "Gid".  Realise this is a just a test, but for future reference in other scenarios.

tomponline commented on lxd/instance_file.go:
```diff
@@ -390,23 +390,29 @@ func instanceFileHead(s *state.State, inst instance.Instance, path string, r *ht
 //	  - in: header
 //	    name: X-LXD-gid
-//	    description: File owner GID
+//	    description: Base 10 integer for the file owner GID
```
Can you split this commit into logical changes:

- Comments and security/validation check improvements.
- Changes specific to the new API extension.

So that we can backport the former and not the latter into the LTS series.

tomponline replied:
It would make sense to me to include shared.ParseLXDFileHeaders change here in a separate commit too, and then have additional commits that update add the new field and shared.ParseLXDFileHeaders to handle the new header and use the new field here.

tomponline commented on client/interfaces.go:
```diff
@@ -693,12 +693,18 @@ type InstanceFileArgs struct {
 	// User id that owns the file
 	UID int64
 
+	ModifyExistingUID bool
```
Shall we use `Set` rather than `Modify` in general in this PR - as its more concise?

tomponline replied:
Also we generally use the naming format of `<Thing><Action>` so would it make more sense for these new vars to be in the form of `UIDSetIfExists`?

MggMuggins replied:
I considered using `Set`, but it doesn't imply that we're changing something that already exists in the same way that `Modify` does: `X-LXD-modify-perm` vs `X-LXD-set-perm` vs `X-LXD-set-existing-perm`. It lets the header itself be a little more concise. Since it's for backwards compat, the header probaby wouldn't make much sense for somebody just browsing the API docs, so any clarity we can provide there is good. I will try and elaborate in the docs as well. I'm happy to change to `Set` if you'd like.

How about `UIDModifyExisting` (& friends)? I think
```go
if UIDModifyExisting { ... }
```
reads better than
```go
if UIDModifyIfExists { ... }
```

tomponline commented on client/lxd_containers.go:
```diff
@@ -809,12 +809,16 @@ func (r *ProtocolLXD) GetContainerFile(containerName string, path string) (io.Re
-	uid, gid, mode, fileType, _ := shared.ParseLXDFileHeaders(resp.Header)
+	headers, err := shared.ParseLXDFileHeaders(resp.Header)
+	if err != nil {
+		return nil, nil, err
```
We should wrap the error here using `return nil, nil, fmt.Errorf("Failed parsing response headers: %w", err)`

tomponline commented on client/lxd_instances.go:
```diff
@@ -1593,6 +1597,24 @@ func (r *ProtocolLXD) CreateInstanceFile(instanceName string, filePath string, a
 		req.Header.Set("X-LXD-write", args.WriteMode)
 	}
 
+	modifyPerm := []string{}
```
you can also just do `var modifyPerm []string` so it'll be initialised to `nil`, you can still append to it and call len on it ok.

tomponline commented on client/lxd_instances.go:
```diff
@@ -1593,6 +1597,24 @@ func (r *ProtocolLXD) CreateInstanceFile(instanceName string, filePath string, a
+		modifyPerm = append(modifyPerm, "mode")
+	}
+
+	if len(modifyPerm) != 0 {
```
Or do we want to just silently fallback to old behaviour on older servers and not include the header at all?

MggMuggins replied:
The server ignores unrecognized headers (I believe I tested sending the new header to a server without the logic to handle it). Somehow I missed `CheckExtension` on my first pass trying to figure out how the list of extensions was used.
My feeling is that silently falling back to the old behavior is probably fine, rather than making this a hard error; we could drop a warning on stderr although that could be obnoxious for library users.

tomponline reviewed (changes requested):
This is look great thanks.

Looking at the PR in more detail now I see there are multiple things going on.
There are improvements to the API docs and security checks, as well as refactoring to accommodate the new API extension (in the form of new structs and functions).

Please could you open a separate PR with the improvements and refactoring changes, that don't include the specifics of the new API extension functionality, and then rebase this one ontop of the new PR so it just has the changes specific to the API extension.

This will allow us to see clearly which changes can be backported and which ones are feature specific. Thanks!

MggMuggins commented on shared/util.go:
```diff
@@ -257,18 +257,47 @@ func LogPath(path ...string) string {
+//     One of `overwrite`, `append`
+//   - `X-LXD-modify-perm`
+//     Comma separated list; 0 or more of `mode`, `uid`, `gid`
+func ParseLXDFileHeaders(headers http.Header) (LXDFileHeaders, error) {
```
In Rust you _have_ to pass ownership of the struct back to the caller, which you can do on the stack like this or with a smart pointer (`Box`); getting used to different patterns :slightly_smiling_face:

MggMuggins commented:
I think I've addressed all your feedback. I'm happy to rebase this onto main after #13231 is merged.

tomponline commented:
Ready for rebase.

MggMuggins commented:
Done.

tomponline commented on shared/util.go:
```diff
@@ -334,11 +340,33 @@ func ParseLXDFileHeaders(headers http.Header) (*LXDFileHeaders, error) {
 		return nil, fmt.Errorf("Invalid file write mode: %q", write)
 	}
 
+	UIDModifyExisting := false
```
just for future reference (no change required) you can also do:

```go
var UIDModifyExisting, GIDModifyExisting, modeModifyExisting bool
```

Also as we're not exporting the vars they should be `uidModifyExisting` (i.e we only capitalise all of them if we are exporting the var and so must capitalise the first letter).

tomponline reviewed (approved):
LGTM thanks!

masnax commented on client/lxd_instances.go:
```diff
@@ -1597,6 +1597,24 @@ func (r *ProtocolLXD) CreateInstanceFile(instanceName string, filePath string, a
+		modifyPerm = append(modifyPerm, "mode")
+	}
+
+	if len(modifyPerm) != 0 && r.CheckExtension("instance_files_modify_permissions") == nil {
```
@MggMuggins Looks like the issue with the permissions not updating after this change is just a typo. The API extension you've added is named `instances_files_modify_permissions`, but here when you check it, `instance` isn't pluralized so it's always skipping the header because it can't find it.

tomponline replied:
Confused how the tests are passing though

masnax replied:
I think the unit test doesn't directly use the client, it instead supplies the headers directly rather than checking the API extension.

MggMuggins replied:
Thanks Max. Fix & tests coming soon
